documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

```diff
@@ -6,283 +6,268 @@ Proposal design for a platform invoke (P/Invoke) mechanism for Silk.NET 3.0.
-```
+**INFORMATIVE:** The Silk.NET team looks back very fondly on the API-as-interfaces scheme present in only the earliest Silk.NET 1.0 previews. The backstory is that AdvancedDLSupport originally required interfaces in order to implement the abstract class, so we added interfaces just for that, but they were removed once they became unnecessary. However, these interfaces are obviously very useful in the advent of C# having advanced mainstream dependency injection. While this is never an explicit target for a low-level, high-speed interoperability library such as Silk.NET, we take pleasure in being able to cater for this use case.
 
+Unless `ThreadLocal` is itself the *static default* configured by the generator user, the thread-local value within `ThreadLocal` should be instantiated with a value factory calling the `Create` method. If `ThreadLocal` is the *static default*, then no value factory is provided and the user must set the value using `MakeCurrent`.
 
-`Conventions` will be used as the primary mechanism for customizing the behaviour of generation, just as `NativeApi` will be used as the primary attribute for this as well. The behaviour of each bit will be described in documentation comments in the Proposed API section.
+The binding class **shall** expose a `static` `Create` method with a `Func<string, nint>` parameter returning an instance of the top-level interface. This **shall** use an instance of the binding class itself.
 
-The Emitter does not do any marshalling. As such, the Emitter **MUST** mandate that every parameter and return type of every function fits the `unmanaged` constraint. For the readers benefit, this can be done using a property on `ITypeSymbol` in Roslyn.
+**INFORMATIVE:** It is undecided whether we want SilkTouch to output the `Create` methods itself or whether we want it in a non-generated partial.
 
-## SilkTouch Overloader
-The Overloader, one of the two final stages whose resources **MUST** be entirely independent of eachother, creates overloads of functions that expose a more user-friendly interface than the function it overloads, and do appropriate marshalling to lower the parameter types used down to the original function's types.
+The binding class **shall** itself implement the top-level interface, using the `getProcAddress` delegate stored from the `Create` method to implement the native calls using a direct function pointer call.
 
-The Overloader **MUST** be able to be used on any function and not be tied to any of the Emitter's constraints.
+The binding class **shall** also contain shorthand functions for calling the static functions contained within the static subinterface using the *static default*.
 
-The Overloader **MUST** be able to be invoked via the SilkTouch CLI and **MAY** be able to be invoked via an incremental Roslyn source generator.
+**INFORMATIVE:** It is undecided whether we want to implement the static subinterface on the binding class itself as well.
```

Comment:
I vote yes to this

Reply:
Yeah fair, just wasn't sure whether we wanted what is essentially a proxy to implement the interface itself, but yeah I don't see why not.

Reply:
```suggestion
The binding class **shall** implement the static subinterface (i.e. to proxy calls to the *static default* for ease of use).
```